### PR TITLE
Add basic reader/writer support for WIG.

### PR DESCRIPTION
**Pull request by r6eve**
This PR adds the feature to support [Wiggle Track Format (WIG)](https://genome.ucsc.edu/goldenpath/help/wiggle.html).

### CODE REVIEW

File: src/cljam/io/wig.clj [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+                   (cond-> wigs
+                     (some? current-track) (conj current-track))
+                   {:format :variable-step
+                    :chrom chrom
+                    :span (as-long span)
+                    :values []}))
+
+          ; declaration line of fixedStep
+          "fixedStep"
+          (let [{:keys [chrom start step span] :or {span 1}} (->> fields rest fields->map)]
```
[the remark this answers is not in the document]
Comment (athos) [reply src/cljam/io/wig.clj]: `span` is also an optional parameter for both `variableStep` and `fixedStep` format, which defaults to 1 if omitted.

File: src/cljam/io/wig.clj [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  positive or negative values."
+  [^String s]
+  (if-let [l (as-long s)]
+    l
+    (as-double s)))
+
+(defn- deserialize-wigs
+  "Parse WIG lines and returns a vector of maps."
+  [lines]
+  (loop [lines lines
```
[the remark this answers is not in the document]
Comment (athos) [reply src/cljam/io/wig.clj]: Nested lazy sequences, in general, are hard to use correctly and I don't think it's a good idea to choose the second option.

As for Option 1. and 3., I think both readers would be useful in different use cases, so I would like both if possible 😄

File: src/cljam/io/wig.clj [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+(defn- serialize-wig
+  "Serialize WIG line into vector of strings."
+  [wig]
+  {:pre [;; There are two options for formatting wiggle data.
+         (or (= (:format wig) :fixed-step) (= (:format wig) :variable-step))
+         ;; These fields are required if the format is fixedStep.
+         (if (= (:format wig) :fixed-step) (and (:chrom wig) (:span wig) (:start wig) (:step wig) (:values wig)) true)
+         ;; These fields are required if the format is variableStep.
+         (if (= (:format wig) :variable-step) (and (:chrom wig) (:span wig) (:values wig)) true)]}
```
Comment (athos) [inline src/cljam/io/wig.clj]: These preconditions are somewhat too complicated. I'd take them away from the function and define a new dedicated predicate to validate the argument WIG data.

Also, since the existence of `span` and `step` parameters are not mandatory (as mentioned in some other comments), drop those conditions from here.

File: src/cljam/io/wig.clj [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+                    :fixed-step
+                    (-> fields first str->wiggle-track-data)
+
+                    (throw (IllegalArgumentException. "Invalid wiggle format")))]
+            (recur (rest lines)
+                   wigs
+                   (update current-track :values #(conj % v)))))))))
+
+(defn- serialize-wig
+  "Serialize WIG line into vector of strings."
```
Comment (athos) [inline src/cljam/io/wig.clj]: The docstring doesn’t seem to describe what the function does, in the sense that it doesn’t return “a vector of strings”.

File: src/cljam/io/wig.clj [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+                      {:start (as-long start)
+                       :value (str->wiggle-track-data value)})
+
+                    :fixed-step
+                    (-> fields first str->wiggle-track-data)
+
+                    (throw (IllegalArgumentException. "Invalid wiggle format")))]
+            (recur (rest lines)
+                   wigs
+                   (update current-track :values #(conj % v)))))))))
```
Comment (athos) [inline src/cljam/io/wig.clj]: FYI: This could be written simply `(update current-track :values conj v)`

File: src/cljam/io/wig.clj (hunk 1 above)
[the remark this answers is not in the document]
Comment (r6eve) [reply src/cljam/io/wig.clj]: Thank you for reviewing. I forgot to set a default value for `step`.

I've already set a default value for `span`. https://github.com/chrovis/cljam/pull/145/files/4f51775479d472b31c6166616675d87a34e2e4f4#diff-3ff272146c2a81163462bf5455386b47R50

File: src/cljam/io/wig.clj (hunk 3 above)
Comment (athos) [reply src/cljam/io/wig.clj]: Ah, let me rephrase that. If you feed default values for those (`span` and `step`) parameters, there's no need to change the conditions at all.

File: src/cljam/io/wig.clj (hunk 1 above)
[the remark this answers is not in the document]
Comment (athos) [reply src/cljam/io/wig.clj]: My bad! Forget all my comments related to `span` and `step` stuff. Sorry to bother you 🙇

File: src/cljam/io/wig.clj (hunk 2 above)
[the remark this answers is not in the document]
Comment (r6eve) [reply src/cljam/io/wig.clj]: Thank you for the detailed comments. I'll choose Option 3. for memory.

Then, I'd like to add a `:format` key to flat map, because a writer cannot determine format without the key. For example, if a writer tries to output this data,
```
({:chr "chr1", :start 1, :end 10, :value 1}
 {:chr "chr1", :start 100, :end 110, :value 2})
```
cannot determine which formats to output.

```
variableStep chrom=chr1 span=10
1 1
100 2
```
```
fixedStep chrom=chr1 start=1 step=100 span=10
1
2
```

**Comment by r6eve** [comment]
Thank you for the careful review. I've fixed it accordingly.

**Comment by r6eve** [comment]
@totakke Thank you for the comments. I fixed tests.

File: src/cljam/io/wig.clj [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+                                                           (not= span 1) (conj (str "span=" span)))
+                                                         (cstr/join \space))
+                                   data-lines (->> track
+                                                   (map (fn [{:keys [start value]}] (cstr/join \space [start value])))
+                                                   (cstr/join \newline))]
+                               (cstr/join \newline [declaration-line data-lines]))
+
+                             :fixed-step
+                             (let [{:keys [chrom start end]} (first track)
+                                   step (- (-> track rest first :start) start)
```
Comment (athos) [inline src/cljam/io/wig.clj]: You can replace `(first (rest xs))` with `(second xs)`.

By the way, I think it would be great it could handle fixed-step tracks with a single data line, for just in case. The current implementation will throw an error in this line if the writer gets fed with such data.

File: src/cljam/io/wig.clj [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+                               {:format :variable-step, :chrom chrom, :start start, :end end, :value value})
+
+                             :fixed-step
+                             (let [{:keys [chrom pre-start span step]} current-track
+                                   start (+ pre-start step)
+                                   end (dec (+ start span))
+                                   value (-> fields first str->wiggle-track-data)]
+                               {:format :fixed-step, :chrom chrom, :start start, :end end, :value value})
+
+                             (throw (IllegalArgumentException. "Invalid wiggle format")))]
```
[the remark this answers is not in the document]
Comment (alumi) [reply src/cljam/io/wig.clj]: Sorry, `format` could be `nil` here.

**Comment by alumi** [comment]
@athos 

> And another thing I'd like to confirm is if there's no need to add writer API for users to specify their custom browser lines (are they out of the scope of WIG format?)

They're okay to be out of the scope, I guess.

We're ignoring "browser lines" because they're very specialized to the UCSC genome browser.

https://github.com/chrovis/cljam/blob/8abbeb9b8829484740b771d269e76d3f05f8f894/src/cljam/io/bed.clj#L143-L149

BED module also ignores the "track definition line" which is not *required* by [the BED file format spec](https://genome.ucsc.edu/FAQ/FAQformat.html#format1).

> BED lines have three required fields and nine additional optional fields.

> Please note that only in custom tracks can the first lines of the file consist of header lines, which begin with the word "browser" or "track" to assist the browser in the display and interpretation of the lines of BED data following the headers.

**Comment by athos** [comment]
@alumi Thanks! That makes sense. I understood they are something like an "implementation-dependent" feature.

**Comment by r6eve** [comment]
Thank you for reviewing. I've updated it.

**Comment by r6eve** [comment]
I'm sorry for missing that. Thank you for pointing it out.

File: src/cljam/io/wig.clj [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+(defn ^WIGWriter writer
+  "Returns an open cljam.io.wig.WIGWriter of f. Should be used inside with-open
+  to ensure the writer is properly closed."
+  [f]
+  (WIGWriter. (cio/writer (util/compressor-output-stream f)) (util/as-url f)))
+
+(def ^:const wig-fields [:track :chr :start :end :value])
+
+(defn- header-or-comment?
+  "Checks if given string is neither a header nor a comment line."
```
Comment (athos) [inline src/cljam/io/wig.clj]: `"Checks if given string is either a header or a comment line."` or more precisely `"Returns true if ..."` would be fine.

**Comment by r6eve** [comment]
@athos Thank you for pointing out. I updated it.

**Comment by alumi** [comment]
Merged. Thanks!!  🎉
